Make mw2pdf respect the -o (--out) CLI parameter

hawkinsw opened the pull request:
This PR accomplishes two main goals:

1. Add types through the mw2pdf codebase as much as humanly possible (and add documentation!). 
2. Make mw2pdf CLI respect the `-o` (`--out`) parameter (Issue #33)

Looking forward to feedback.

slifty commented on services/api/mw2pdf/src/utils/validation.ts:
```diff
@@ -30,16 +30,24 @@ const passthroughParametersSchema = Joi.object()
     JSON.parse(passthroughParameters),
     passthroughParametersSchema,
-  )
+  );
```
(This comment applies to a bunch of lines; I won't repeat it on all of them)

Right now this code base doesn't use semi's; we don't have linting so this isn't explicit, but lets make sure to follow that pattern for now.

hawkinsw replied:
You got it!! I come from the opposite "world" where our linter made sure that we always had semis. I will undo that immediately!

slifty replied:
Yeah the semis vs not is age old and I'm by no means suggesting they shouldn't ever be used -- just this project actually didn't have them before I came in (I've tended to adopt the no semis personally but don't begrudge them)

slifty commented on services/api/mw2pdf/src/utils/url.ts:
```diff
@@ -1,11 +1,13 @@
+ * @returns `url` with updated parameters.
+ */
+export function updateUrlParameters(url: URL, parameters: Array<[key: string, value: string]>) {
+  parameters.forEach(([key, value]) => url.searchParams.set(key, String(value)));
```
I realize the whitespace here is just a style choice (I did explicitly break over multiple lines for readability, but may have gone overboard).

At any rate, it would be good to relegate whitespace changes to their own commit -- and ideally we would do it as a single "style / linting" PR at a later date when we set up eslint.

hawkinsw replied:
I *think* that I got all the places!

slifty replied:
So ironically this one wasn't but ALSO!

I am realizing this was actually changed, not just whitespace.  The change introduces a bug -- parameters isn't an Array, but an object.  The original implementation used Object.entries to get the keys of the parameters object and then iterate through them.

hawkinsw replied:
This should now be fixed. As we talked about in chat, "stupid TS not being strict enough!"

slifty commented on services/api/mw2pdf/src/utils/mediaWiki.ts:
```diff
@@ -18,29 +27,47 @@ export async function makeLoginTokenRequest(apiUrl) {
+  password: string = "",
+  loginToken: string = "",
+  cookies: Array<ParsedCookie> = []): Promise<Response> {
+
```
rogue newline should be removed

hawkinsw replied:
Gone!

slifty commented on services/api/mw2pdf/src/utils/files.ts:
```diff
@@ -1,5 +1,11 @@
-import fs from 'fs'
+import { Pdf } from "src/classes/Pdf";
```
I believe we're only using this as a type so this should be `import type` (also we really should set up eslint which would catch this)

hawkinsw replied:
I never used this feature of TS before!! So cool!

slifty commented on services/api/mw2pdf/src/index.ts:
```diff
@@ -15,21 +24,31 @@ program
   .option('--mwPassword <string>', 'The password to log in with', '')
   .option('--passthroughParameters <string>', 'a json encoded string containing data to pass via querystring with each request', '')
-  .action(async (urls, options) => {
+  .action(async (urlStrings: Array<string>, options: CliOptions) => {
```
omg I am so happy about your change from `urls` to `urlString` so we can avoid `processedUrls` below.

slifty commented on services/api/mw2pdf/src/index.ts:
```diff
@@ -15,21 +24,31 @@ program
+        } catch (e) {
+          throw new Error("Invalid URL");
+        }
+      });
```
I think this is indented by one extra space

hawkinsw replied:
It is!

slifty commented on services/api/mw2pdf/src/index.ts:
```diff
@@ -15,21 +24,31 @@ program
+      }
     }
 
+    let urls: Array<URL> = urlStrings.map((urlString: string) => {
```
Could we put this into a named utility method -- `generateUrls` or something.

hawkinsw replied:
Your wish ... has been granted. :-)

slifty commented on services/api/mw2pdf/src/index.ts:
```diff
@@ -3,7 +3,16 @@ import Joi from 'joi'
-import { MediaWikiSession } from './classes/MediaWikiSession'
+import { MediaWikiSession, MwPdfOptions } from './classes/MediaWikiSession'
+
+interface CliOptions {
```
`CliOptions` feels like an awkward name but I can't nail down why...  What about `CommandOptions` -- since the fact that it's CLI is kind of abstracted away by `commander` anyway.

hawkinsw replied:
Great call!! Switching!

slifty commented on services/api/mw2pdf/src/classes/PdfFactory.ts:
```diff
@@ -93,8 +86,9 @@ export class PdfGenerator {
       },
-      [],
-    )
+      Promise.resolve(new Array<Toc>()), // Start the reduction with an empty slate.
```
The fact that reduce as a concept is so complicated that a standard parameter warrants a comment to explain is just further proof that reduce should be avoided at all costs.

(This isn't a slam on the comment -- I approve)

hawkinsw replied:
This one actually required me to Google :-) So, yes, I am in strong agreement!

slifty commented on services/api/mw2pdf/src/classes/PdfFactory.ts:
```diff
@@ -1,51 +1,34 @@
 const __dirname = path.dirname(__filename);
 
-export class PdfGenerator {
+export class PdfFactory {
```
MY DESIGN PATTERNS MIND IS BLOWN

hawkinsw replied:
LOL! Gang of Four!

slifty commented on services/api/mw2pdf/src/classes/Pdf.ts:
```diff
@@ -1,36 +1,107 @@
+    const data = fs.readFileSync(this.filename);
+    const pdf = await (getDocument(data).promise);
+    const page = await pdf.getPage(1);
+    return new PdfInfo(pdf.numPages, new PdfSize(page.view[2] as number, page.view[3] as number));
```
This is a lot in one line, could we split it up a little?

hawkinsw replied:
Absolutely!!

slifty commented on services/api/mw2pdf/src/classes/Pdf.ts:
```diff
@@ -1,36 +1,107 @@
+/**
+ * A data structure to hold information about the size of a PDF
+ */
+export class PdfSize {
```
Once we add linting I know we're gonna get slapped with the "too many classes in one file" warning.

That said I think most of these are actually just [`Interfaces`](https://www.typescriptlang.org/docs/handbook/interfaces.html) instead of classes, so we should be able to just make that swap and won't violate the one-class-to-a-file goal.

slifty commented on services/api/mw2pdf/src/classes/MediaWikiSession.ts:
```diff
@@ -1,131 +1,219 @@
+     */
+    const pagePdfs: Array<Pdf> = await Promise.all(
+      urls.map(async (url) => {
+        const pagePdfOptions = new MwPdfOptions("No Title", null, options.workDirectory, MwPdfMakeTitlePage);
```
We've been using single quotes `'` for strings rather than double quotes; lets keep that standard (again, eslint would solve this)

hawkinsw replied:
Done!

slifty commented on services/api/mw2pdf/src/classes/MediaWikiSession.ts:
```diff
@@ -1,131 +1,219 @@
+     * End puppeteer
+     */
+
+
```
Too much whitespace here

slifty commented on services/api/mw2pdf/src/classes/MediaWikiSession.ts:
```diff
@@ -1,131 +1,219 @@
+    // print the visible page into a PDF
+    await page.pdf({ path: pagePdf.filename, format: 'A4' });
+
+    // close the browser
```
I don't think we need this comment since it's just a repeat of the below line (`browser.close` vs `close the browser`)

hawkinsw replied:
As if it never appeared :-)

slifty commented on services/api/mw2pdf/src/classes/MediaWikiSession.ts:
```diff
@@ -1,131 +1,219 @@
+    // that all the login and authentication we did before will be valid.
+    await page.setCookie(...this.getCookies(domain));
+
+    // Now, go to the page!
```
There are some other comments that are redundant (which is a good problem to have -- it means we named our variables / methods very well!

hawkinsw replied:
I removed the ones that I thought were redundant. Let me know what you think!

slifty commented on services/api/mw2pdf/src/classes/MediaWikiSession.ts:
```diff
@@ -1,131 +1,219 @@
     const browser = await puppeteer.launch(
       { args: ['--no-sandbox', '--disable-setuid-sandbox'] }
-    )
-    const page = await browser.newPage()
```
I do think that overall adding comments at the expense of spreading these 4 lines over much more actually makes it a little less legible.  Just a sanity check: did you find this block un-intuitive as written?

hawkinsw replied:
I did, actually. At first it just appeared like a wall of instructions and I wasn't sure exactly what was going on. So, I just blew them apart and added comments mostly for myself. If they are not helpful, I am happy to remove them!

slifty replied:
Totally legit -- thanks for bearing with me on the comment stuff, I appreciate that it's nuanced and there is no clear "right" so sorry if it felt like annoying bike shedding.

Thanks for helping make this clear.

slifty commented on services/api/mw2pdf/src/classes/MediaWikiSession.ts:
```diff
@@ -1,131 +1,219 @@
-  async makePdf(url, tmpDirectory = './', includeTitlePage = false) {
-    const pdfGenerator = new PdfGenerator(tmpDirectory)
+  async makePdf(url: URL, options: MwPdfOptions): Promise<Pdf> {
+    // Extract the domain (as _domain_) from the url.
```
I'm not sure what this comment means (beyond what the code itself said already)

hawkinsw replied:
Voila! Gone.

slifty commented on services/api/mw2pdf/src/classes/MediaWikiSession.ts:
```diff
@@ -1,131 +1,219 @@
+ * A class to hold the options that control creation of a PDF (booklet)
+ * from a MediaWiki site.
+ */
+export class MwPdfOptions {
```
Can this be turned into an Interface instead of a Class?

hawkinsw replied:
Absolutely! And it is!

slifty reviewed (changes requested):
Really excited for the TS changes here and I also think the refactoring is great!

Some changes requested are:

1. Make sure we don't add semicolons
2. Avoid pure whitespace changes
3. I'd like to avoid introducing comments that repeat (or even explain) what the code is doing [here is a fun article about types of comments](https://www.red-gate.com/simple-talk/opinion/opinion-pieces/fighting-evil-code-comments-comments/)
4. There are a few classes that I think should be Interfaces
5. Fix a few introduced linting errors before they come (mostly extra whitespace in a few spots)

slifty commented on services/api/mw2pdf/src/classes/PdfFactory.ts:
```diff
@@ -1,51 +1,34 @@
+    const merger = new PdfMerger();
+    pdfs.forEach(pdf => merger.add(pdf.filename));
+    await merger.save(outPdf.filename);
+    return outPdf;
```
semicolons in this method (some which created changed lines that aren't related to the PR)

hawkinsw replied:
Took out all the semis in this file as well!

slifty commented on services/api/mw2pdf/src/utils/cookies.ts:
```diff
@@ -1,13 +1,37 @@
+ * @param cookies An array of [[ParsedCookie]]s from which to create a string.
+ */
+export function toCookieString(cookies: Array<ParsedCookie>): string {
+  return cookies.map(cookie => `${cookie.name}=${cookie.value}`).join(';');
```
lurking semicolon change here

hawkinsw replied:
I took all the semis out of this file! I think that was the Right Thing (TM) to do.

slifty reviewed (changes requested):
Yay!

A few lurking semicolon changes we should revert and one bug to be fixed.  Looking really good.

hawkinsw committed:
Add types and documentation

Introduce types as much as humanly possible throughout the mw2pdf
codebase. Also, add as much documentation as possible to the mw2pdf
codebase.

This commit pays off some technical debt.

hawkinsw committed:
Fix mw2pdf CLI that does not obey the -o parameter

mw2pdf does not currently obey the -o (--out) parameter. This patch
fixes that.

Issue #33.

slifty reviewed (approved):
Hooray!  I think this is it!  Thanks for all the iteration.
